### ISSUE TITLE
[rllib] provides a flattened observation, instead of the original, for custom_loss()

### ISSUE THREAD

**Issue by ArturNiederfahrenhorst** [ray-project/ray#7896]
The following code, shows a case where the `input_dict` arriving in the method `custom loss()` does not contain the original observation dictionary, but rather a flattened version of it. This is impractical, as we can not see what part of it corresponds to what key of the original observation dictionary. This is not an issue when working with the `ray.rllib.models`, since they accept the `input_dict` directly. But when building custom tensorflow models, the dictionary comes in very handy.

```
import ray
import gym
import ray.tune as tune
import tensorflow as tf
from ray.rllib.models import ModelCatalog
from ray.rllib.models.tf.tf_modelv2 import TFModelV2


class DummyEnv(gym.Env):
	def __init__(self, env_config):
		self.action_space = gym.spaces.Box(low=-1, high=1, shape=(1,))
		self.observation_space = gym.spaces.Dict({
			'obs1': gym.spaces.Box(low=-1, high=1, shape=(1,)),
			'obs2': gym.spaces.Box(low=-1, high=1, shape=(1,))
		})

	def reset(self):
		pass

	def step(self, action):
		observation = {'obs1': 0,
		               'obs2': 1}

		return observation, 0, False, {}


class DummyModel(TFModelV2):
	def __init__(self, obs_space, action_space, num_outputs, model_config, name):
		super(DummyModel, self).__init__(obs_space, action_space, num_outputs, model_config, name)

		input1 = tf.keras.layers.Input(shape=[1])
		input2 = tf.keras.layers.Input(shape=[1])
		output = tf.keras.layers.Concatenate()([input1, input2])
		self.model = tf.keras.models.Model([input1, input2], [output])

		self.register_variables(self.model.variables)

	def forward(self, input_dict, state, seq_lens):
		return self.model([input_dict['obs']['obs1'], input_dict['obs']['obs2']]), []

	def value_function(self):
		return tf.fill([1], 0.0, name=None)

	def custom_loss(self, policy_loss, loss_inputs):
		print(loss_inputs['obs'])  # Critical part
		return policy_loss


if __name__ == '__main__':
	ray.init()
	ModelCatalog.register_custom_model("dummy_model", DummyModel)
	tune.run(
		"PPO",
		config={
			"env": DummyEnv,
			"model": {
				"custom_model": "dummy_model",
			},
		},
	)
```

I propose that the loss_inputs should come with the full obersavtion dictionary.

**Comment by 51616** [comment]
Is there any update on this feature?

**Comment by ArturNiederfahrenhorst** [comment]
> Is there any update on this feature? or is there any way we could know which dimension is from which key?
> 
> I'm trying to add `custom_loss` to my model with auto-encoder loss, but the target observation is from the other agent.
> If I can't access the nested dict here (which includes other agent's observation), how can I do a work around for this problem?
> 
> Edit: Solved
> There's a function that restore the original dimensions
> 
> ```
> from ray.rllib.models.modelv2 import restore_original_dimensions
> obs = restore_original_dimensions(loss_inputs["obs"], self.obs_space)
> ```

Thanks for posting this. I still think that, for most people, it would just be convenient to have the dictionary at hand. Especially because `custom_loss()` is very exposed to newer users.

**Comment by rallen10** [comment]
I have a related question but it has to do with callbacks instead of custom models and custom_loss. If I want to modify observations from a batch (e.g. for [centralized critic](https://github.com/ray-project/ray/blob/releases/0.8.7/rllib/examples/centralized_critic_2.py#L33)), how do I know which dimensions of `SampleBatch.CUR_OBS` correspond to which keys in the Dict observation space? I tried the `restore_original_dimensions` suggested by @51616 but it just returned another array, not the reconstructed dictionary.

**Comment by sven1977** [comment]
@rallen10 Did you pass in the correct original space into `restore_original_dimensions([obs_flat], [original Dict space], framework="tf|torch")`.

**Comment by rallen10** [comment]
@sven1977 thank you for the follow-up. I can't remember how I worked around that issue, but I'll know what to try in the future

**Comment by loicsacre** [comment]
RLLib also provides a flattened observation for the `compute_actions` method of the `Policy` class. The only way I found to use the method `restore_original_dimensions` in `compute_actions` is to do the following:

```
import torch
obs = np.expand_dims(obs_batch[0], axis=0)
obs = torch.from_numpy(obs)
original_obs = restore_original_dimensions(obs, self.observation_space, "torch")
```

Is there a more convenient way to do so (working with np.array) ?

**Comment by ArturNiederfahrenhorst** [comment]
Closing this because of the new RLModules API does not have this problem.
User can decide to not use a preprocessor and all forward methods of the RL Module will receive unflattened data.

**Comment by edcxan** [comment]
> Closing this because of the new RLModules API does not have this problem. User can decide to not use a preprocessor and all forward methods of the RL Module will receive unflattened data.

How is it done in RL Module?

**Comment by omsrisagar** [comment]
Yes, @ArturNiederfahrenhorst , could you please provide a reference/few lines of code on how to use that preprocessor (to avoid flattening)? Thank you for your help!
